src/Microsoft.Azure.WebJobs.Host/Singleton/BlobLeaseDistributedLockManager.cs

```diff
@@ -0,0 +1,404 @@
+            await ReleaseLeaseAsync(singletonLockHandle.Blob, singletonLockHandle.LeaseId, cancellationToken);
+        }
+
+        public async virtual Task<string> GetLockOwnerAsync(string account, string lockId, CancellationToken cancellationToken)
+        {
+            IStorageBlobDirectory lockDirectory = GetLockDirectory(account);
+            IStorageBlockBlob lockBlob = lockDirectory.GetBlockBlobReference(lockId);
+
+            await ReadLeaseBlobMetadata(lockBlob, cancellationToken);
+
+            // if the lease is Available, then there is no current owner
+            // (any existing owner value is the last owner that held the lease)
+            if (lockBlob.Properties.LeaseState == LeaseState.Available &&
+                lockBlob.Properties.LeaseStatus == LeaseStatus.Unlocked)
+            {
+                return null;
+            }
+
+            string owner = string.Empty;
+            lockBlob.Metadata.TryGetValue(FunctionInstanceMetadataKey, out owner);
+
+            return owner;
+        }
+
+        public async Task<IDistributedLock> TryLockAsync(
+            string account,
+            string lockId,
+            string lockOwnerId,
+            TimeSpan lockPeriod,
+            CancellationToken cancellationToken)
+        {
+            IStorageBlobDirectory lockDirectory = GetLockDirectory(account);
+            IStorageBlockBlob lockBlob = lockDirectory.GetBlockBlobReference(lockId);
+            string leaseId = await TryAcquireLeaseAsync(lockBlob, lockPeriod, cancellationToken);
+
+            if (string.IsNullOrEmpty(leaseId))
+            {
+                return null;
+            }
+
+            if (!string.IsNullOrEmpty(lockOwnerId))
+            {
+                await WriteLeaseBlobMetadata(lockBlob, leaseId, lockOwnerId, cancellationToken);
+            }
+
+            SingletonLockHandle lockHandle = new SingletonLockHandle(leaseId, lockId, lockBlob, lockPeriod);
+                                    
+            return lockHandle;
+        }
+
+        internal IStorageBlobDirectory GetLockDirectory(string accountName)
+        {
+            if (string.IsNullOrEmpty(accountName))
+            {
+                accountName = ConnectionStringNames.Storage;
+            }
+
+            IStorageBlobDirectory storageDirectory = null;
+            if (!_lockDirectoryMap.TryGetValue(accountName, out storageDirectory))
+            {
+                Task<IStorageAccount> task = _accountProvider.GetStorageAccountAsync(accountName, CancellationToken.None);
+                IStorageAccount storageAccount = task.Result;
+                // singleton requires block blobs, cannot be premium
+                storageAccount.AssertTypeOneOf(StorageAccountType.GeneralPurpose, StorageAccountType.BlobOnly);
+                IStorageBlobClient blobClient = storageAccount.CreateBlobClient();
+                storageDirectory = blobClient.GetContainerReference(HostContainerNames.Hosts)
```

Comment:
>HostContainerNames.Hosts [](start = 68, length = 24)

It would be better to pass the container, directory and blob names as parameters to the LockAsync() method instead of assuming names here. Besides, I think the BlobLeaseManager defines its own container name (azure-webjobs-hosts) and needs no directory.